src/protocol/boolean/bitwise_lt.rs

```diff
@@ -0,0 +1,258 @@
+use super::prefix_or::PrefixOr;
+use super::BitOpStep;
+use crate::error::BoxError;
+use crate::ff::BinaryField;
+use crate::protocol::{context::ProtocolContext, mul::SecureMul, RecordId};
+use crate::secret_sharing::Replicated;
+use futures::future::try_join_all;
+use std::iter::{repeat, zip};
+
+/// This is an implementation of Bitwise Less-Than on bitwise-shared numbers.
+///
+/// `BitwiseLessThan` takes inputs `[a]_B = ([a_1]_p,...,[a_l]_p)` where
+/// `a1,...,a_l ∈ {0,1} ⊆ F_p` and `[b]_B = ([b_1]_p,...,[b_l]_p)` where
+/// `b1,...,b_l ∈ {0,1} ⊆ F_p`, then computes `h ∈ {0, 1} <- a <? b` where
+/// `h = 1` iff `a` is less than `b`.
+///
+/// Note that `[a]_B` can be converted to `[a]_p` by `Σ (2^i * a_i), i=0..l`. In
+/// other words, if comparing two integers, the protocol expects inputs to be in
+/// the little-endian; the least-significant byte at the smallest address (0'th
+/// element).
+///
+/// 5.3 Bitwise Less-Than
+/// "Unconditionally Secure Constant-Rounds Multi-party Computation for Equality, Comparison, Bits, and Exponentiation"
+/// I. Damgård et al.
+pub struct BitwiseLessThan<'a, B: BinaryField> {
```

Comment:
> As this comment says: the inputs are all elements of a larger prime field: ⊆ F_p. This implementation is assuming a binary field: B: BinaryField.

The paper says `a1,...,a_l ∈ {0,1} ⊆ F_p`. I interpreted this as; `{0, 1} ⊆ F_p` then `{0,1} ∈ Z_2 ⊆ F_p` is also true. Therefore, I thought I could use `BinaryField` to simplify algebraic operations rather than using `Field` with `mod 2` operations. If so, then the code actually matches the comment since `BinaryField ⊆ Field`. Maybe I'm missing to see cases where using `BinaryField` becomes a problem?

> All of this work is only necessary if you're operating in a larger field. Otherwise there are much, much simpler ways to compute "less than".

Yes. we want to check `a_p <? b_p`, and not `[a]_p <? [b]_p` or `[a]_B <? [b]_B`.

Let me clarify for to make sure my understanding is correct. The inputs to `BitwiseLessThan` are bitwise-shared secrets `[a]_B` and `[b]_B`. That is, we have a secret integer `a ⊆ F_p`, and transform it into bits `a_B = (a_0,...a_l) where a_i {0,1} ⊆ Z_2`. Then we XOR secret-share each bit `a_i` to `[a_i]` (equivalent to additive sharing in `Z_2`), which forms `[a]_B = ([a_1],...[a_l])`. Hence, we cannot reconstruct an additive sharing of `[a]_p` from `[a]_B`.

So, this comment was incorrect:

> Note that `[a]_B` can be converted to `[a]_p` by `Σ (2^i * a_i), i=0..l`

It should be:

> Note that `a_B` can be converted to `a_p` by `Σ (2^i * a_i), i=0..l`.

My intention here was to just say that a binary representation of `a` needs to be in little-endian format.